server/noderelay/cmd/sourcenode/main.go

```diff
@@ -0,0 +1,276 @@
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	killChan := make(chan os.Signal, 1)
+	signal.Notify(killChan, os.Interrupt)
+	go func() {
+		<-killChan
+		fmt.Println("Shutting down...")
+		cancel()
+	}()
+
+	// Keep track of some basic stats.
+	var stats struct {
+		requests uint32
+		errors   uint32
+		received uint64
+		sent     uint64
+	}
+
+	// Periodically print the node usage statistics.
+	go func() {
+		start := time.Now()
+		for {
+			select {
+			case <-time.After(time.Minute * 10):
+			case <-ctx.Done():
+				return
+			}
+			log.Infof("%d requests, %.4g MB received, %.4g MB sent, %d errors in %s",
+				atomic.LoadUint32(&stats.requests), float64(atomic.LoadUint64(&stats.received))/1e6,
+				float64(atomic.LoadUint64(&stats.sent))/1e6, atomic.LoadUint32(&stats.errors),
+				time.Since(start))
+		}
+	}()
+
+	localNodeURL := "http://127.0.0.1" + ":" + port
+	httpClient := http.DefaultClient
+	if localNodeCert != "" {
+		localNodeURL = "https://127.0.0.1" + ":" + port
+		pem, err := os.ReadFile(localNodeCert)
+		if err != nil {
+			return err
+		}
+
+		uri, err := url.Parse(localNodeURL)
+		if err != nil {
+			return fmt.Errorf("error parsing URL: %v", err)
+		}
+
+		pool := x509.NewCertPool()
+		if ok := pool.AppendCertsFromPEM(pem); !ok {
+			return fmt.Errorf("invalid certificate file: %v", localNodeCert)
+		}
+		tlsConfig := &tls.Config{
+			RootCAs:    pool,
+			ServerName: uri.Hostname(),
+		}
+
+		httpClient = &http.Client{
+			Transport: &http.Transport{
+				TLSClientConfig: tlsConfig,
+			},
+		}
+	}
+
+	// Define this now, so we can create a ReconnectSync function.
+	var cl comms.WsConn
+
+	registerOrReregister := func() {
+		if err := cl.SendRaw(registration); err != nil {
+			log.Errorf("Error sending registration message: %v", err)
+		}
+	}
+
+	cl, err = comms.NewWsConn(&comms.WsCfg{
+		URL:      "wss://" + nexusAddr,
+		PingWait: noderelay.PingPeriod * 2,
+		Cert:     certB,
+		// On a disconnect, wsConn will attempt to reconnect immediately. If
+		// the first attempt is unsuccessful, it will wait 5 seconds for next
+		// success, then 10, 15 ... up to a minute.
+		// We'll just send the registration on reconnect.
+		ReconnectSync:    registerOrReregister,
+		ConnectEventFunc: func(s comms.ConnectionStatus) {},
+		Logger:           dex.StdOutLogger("CL", dex.LevelDebug),
+		RawHandler: func(b []byte) {
+			atomic.AddUint64(&stats.received, uint64(len(b)))
+			atomic.AddUint32(&stats.requests, 1)
+			// Request received from server.
+			var msg noderelay.RelayedMessage
+			if err := json.Unmarshal(b, &msg); err != nil {
+				atomic.AddUint32(&stats.errors, 1)
+				log.Errorf("json unmarshal error: %v", err)
+				return
+			}
+			// Prepare mirrored request for local service.
+			ctx, cancel := context.WithTimeout(ctx, time.Second*30)
+			defer cancel()
+			req, err := http.NewRequestWithContext(ctx, http.MethodPost, localNodeURL, bytes.NewReader(msg.Body))
```

Comment:
Will there definitely never be a GET request? I think it's better to also send the method for safety.